Drone/drone_base.py

```diff
@@ -0,0 +1,361 @@
+    connected: Boolean
+        True if the drone is connected to Ardupilot
+    """
+    __metaclass__ = abc.ABCMeta
+
+    cnt = 0
+
+    def __init__(self):
+        self.vehicle = None
+        self.devices = []
+        self.connected = False
+        self.logger = logging.getLogger(__name__)
+
+    def connect(self, isInSimulator):
+        """
+        Attempts to connect to Ardupilot
+
+        Parameters
+        ----------
+        isInSimulator: Boolean
+            True if the program is intended to run in the Gazebo simulator
+            and False if the program is being run on the real-life drone
+
+        Precondition:
+        ----------
+        None
+
+        Postcondition:
+        ----------
+        Upon a successful connection, self.connected is set to True.
+        The drone is now suitable to be armed.
+
+        Returns:
+        ----------
+        None
+        """
+        if isInSimulator:
+            self.vehicle = dronekit.connect(c.CONNECTION_STRING_SIMULATOR, wait_ready=True, 
+                status_printer=None)
+            self.logger.info(threading.current_thread().name + ": Connecting to the simulated ardupilot")
+        else:
+            self.vehicle = dronekit.connect(c.CONNECTION_STRING_REAL, wait_ready=True, 
+                status_printer=None)
+            self.logger.info(threading.current_thread().name + ": Connecting to the real-life ardupilot")
+
+        if self.vehicle is not None:
+            self.connected = True
+
+    def altitude(self):
+        """
+        Simple wrapper function to retrieve the distance being read by 
+        the drone's rangefinder
+
+        Returns:
+        ----------
+        double
+            The distance from the ground
+        """
+        return self.vehicle.rangefinder.distance
+
+    def altitude_failsafe(self):
+        """
+        Simple wrapper function to retrieve the distance being read by 
+        the drone's barometer
+
+        Returns:
+        ----------
+        double
+            The distance from the ground
+        """
+        return self.vehicle.location.global_relative_frame.alt
+
+    def arm(self, timeout = 60):
+        """
+        Attempts to arm the drone for flight
+
+        Parameters
+        ----------
+        timeout: Integer (optional)
+            The duration in seconds that arming should be attempted before
+            giving up and returning
+
+        Precondition:
+        ----------
+        self.connect has been called and succeeded.
+
+        Postcondition:
+        ----------
+        Upon successful arm, self.vehicle.armed is True, and false otherwise.
+        The drone is now suitable to take off.
+
+        Returns:
+        ----------
+        None
+        """
+        start = time.time()
+        self.vehicle.mode = VehicleMode(c.GUIDED_MODE)
+
+        self.logger.info(threading.current_thread().name + ": Arming")
+        start = time.time()
```

Comment:
Consider making a timer class for this purpose if this pattern is common throughout your code. Here's an example of one that I've created:

```Python
from timeit import default_timer as timer

class Timer():
    def __init__(self):
        self._event_threads = []
        self._stop_event_map = {}
        self._thread_map = {}
        self._lock = threading.Lock()
        self._num_threads_lock = threading.Lock()

        self.reset()

    def _create_event_thread(self, closure, name):
        stop_event = threading.Event()
        self._stop_event_map[name] = stop_event
        thread = threading.Thread(target=closure, daemon=True, args=(stop_event,))
        self._thread_map[name] = thread
        self._event_threads.append(thread)
        thread.start()

    @property
    def num_threads(self):
        with self._num_threads_lock as lock:
            return len(self._event_threads)

    def add_callback(self, name, when_to_call, callback, recurring=False):
        # when_to_call: call when x seconds are elapsed
        if name in self._stop_event_map:
            raise ValueError('The `name` parameter cannot be assigned the duplicate name: "{}"!'.format(name))

        if when_to_call <= self.elapsed:
            callback()
            return

        def handle_event(stop_event):
            while recurring and not stop_event.is_set():
                time.sleep(when_to_call-self.elapsed)
                event_handled = False
                while ((not event_handled) or recurring) and not stop_event.is_set():
                    if when_to_call <= self.elapsed:
                        callback()
                        event_handled = True
                        break

                    time.sleep(1e-3)

                if self.num_threads == 1 and recurring:
                    self.reset()

        self._create_event_thread(handle_event, name)

    def stop_callback(self, name):
        self._stop_event_map[name].set()

    def shutdown(self):
        for name, stop_event in self._stop_event_map.items():
            stop_event.set()
            self._thread_map[name].join()

    @property
    def elapsed(self):
        with self._lock as lock:
            return timer() - self._start

    def reset(self):
        with self._lock as lock:
            self._start = timer()
```

Reply:
Also, note the documentation for time is:
```Python
Return the time in seconds since the epoch as a floating point number. Note that even though the time is always returned as a floating point number, not all systems provide time with a better precision than 1 second. While this function normally returns non-decreasing values, it can return a lower value than a previous call if the system clock has been set back between the two calls.```

Reply:
I have defined the suggested Timer class in the Utilities/timer.py.